crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Mauro Romito
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use eyeball_im::{ObservableVector, VectorDiff, VectorSubscriber};
+use futures_core::Stream;
+use ruma::{
+    api::client::directory::get_public_rooms_filtered::v3::Request as PublicRoomsFilterRequest,
+    directory::{Filter, PublicRoomJoinRule},
+    OwnedMxcUri, OwnedRoomAliasId, OwnedRoomId,
+};
+
+use crate::{Client, Result};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RoomDescription {
+    pub room_id: OwnedRoomId,
+    pub name: Option<String>,
+    pub topic: Option<String>,
+    pub alias: Option<OwnedRoomAliasId>,
+    pub avatar_url: Option<OwnedMxcUri>,
+    pub join_rule: PublicRoomJoinRule,
+    pub is_world_readable: bool,
+    pub joined_members: u64,
+}
+
+pub struct RoomDirectorySearch {
+    batch_size: u32,
+    filter: Option<String>,
+    next_token: Option<String>,
+    client: Client,
+    results: ObservableVector<RoomDescription>,
+    is_at_last_page: bool,
+}
+
+impl RoomDirectorySearch {
+    pub fn new(client: Client) -> Self {
+        Self {
+            batch_size: 0,
+            filter: None,
+            next_token: None,
+            client,
+            results: ObservableVector::new(),
+            is_at_last_page: false,
+        }
+    }
+
+    pub async fn search(&mut self, filter: Option<String>, batch_size: u32) -> Result<()> {
```

Comment:
This design works because `RoomDirectorySearch` **is not** thread-safe. Calling `search` while `next_page` is running would be a design flow. I'm all good for the current design, but we must ensure that `RoomDirectorySearch` is never used concurrently. Maybe it is important enough to be precised in the documentation?